### PR TITLE
feat(debug_helper): Adding debug helper when a get call fails

### PR DESCRIPTION
**Pull request by antoaravinth**
Fix for https://github.com/kentcdodds/react-testing-library/issues/31

@kentcdodds Let me know if things are good to go!

### CODE REVIEW

File: src/queries.js [hunk 1]
```diff
@@ -1,4 +1,6 @@
 import {matches} from './matches'
+import prettyFormat from 'pretty-format' // eslint-disable-line
```
Comment (sompylasar) [inline src/queries.js]: Why eslint-disable-line here?

File: src/queries.js [hunk 2]
```diff
@@ -129,11 +165,40 @@ function queryByAltText(container, alt) {
...
+      }),
+    )
+    throw new Error(
+      `Unable to find an element with the alt text: ${alt} \nHere is the state of your container: \n${htmlElement}`,
+    )
   }
   return el
 }
 
+function checkDebugLimit(debugContent) {
```
Comment (sompylasar) [inline src/queries.js]: The name is confusong: it shouldn't be "check" (would return boolean), it is more of a "trimToDebugLimit".

File: src/queries.js [hunk 3]
```diff
@@ -129,11 +165,40 @@ function queryByAltText(container, alt) {
...
+  const limit = process.env.DEBUG_PRINT_LIMIT
```
Comment (sompylasar) [inline src/queries.js]: What if the limit needed in browser where there's no env, e.g. via Cypress?

I suggest exporting a mutable "defaultOptions" object from the module instead, like Node's util.inspect.defaultOptions.

File: src/queries.js [hunk 4]
```diff
@@ -129,11 +165,40 @@ function queryByAltText(container, alt) {
...
+    if (limit <= contentLength)
+      // there are more chars we have stripped, just show to the user
+      debugContent += '\n . . .'
+  } else {
+    // default limit is 7000
+    const contentLength = debugContent.length
+    debugContent = debugContent.slice(0, 7000)
+    if (contentLength >= 7000)
+      // there are more chars we have stripped, just show to the user
+      debugContent += '\n . . .'
```
Comment (sompylasar) [inline src/queries.js]: The above "if" and "else" are essentially copy-paste except `limit` is replaced with a default value.

I suggest pulling the "magic number" into a "defaultOptions" object and use it like so:
```
    const limit = defaultOptions.limit;
    // OR:  const limit = defaultOptions.limit || 7000;
    const contentLength = debugContent.length
    debugContent = debugContent.slice(0, limit)
    if (limit <= contentLength)
      // there are more chars we have stripped, just show to the user
      debugContent += '\n . . .'
```

Comment (antoaravinth) [reply src/queries.js]: Yes, I agree. Can be improved. This is just a rough draft, I have to clean up few things. Thanks for your comments.

File: src/queries.js (hunk 3 above)
Comment (antoaravinth) [reply src/queries.js]: Yes, thats a good point.

File: src/queries.js (hunk 1 above)
Comment (antoaravinth) [reply src/queries.js]: So basically `pretty-format` is not part of the package.json, dependencies but its part of `jest` I believe. Hence I have added the `eslint-disable-line`, yes, I need to clean up a bit by adding the exact eslint-ignore.

Comment (sompylasar) [reply src/queries.js]: But what if I don't use jest with dom-testing-library?

Comment (antoaravinth) [reply src/queries.js]: Sorry, `pretty-format` is from `jest-matcher-utils` (https://github.com/facebook/jest/blob/master/packages/jest-matcher-utils/package.json#L14) which is our dependencies. Sorry for the rushed comment, saying it as `jest`!

**Comment by antoaravinth** [comment]
@kentcdodds Hey Kent.. @sompylasar suggested the following:

> What if the limit needed in browser where there's no env, e.g. via Cypress?
> 
> I suggest exporting a mutable "defaultOptions" object from the module instead, like Node's util.inspect.defaultOptions.

w.r.t to process env for `DEBUG_PRINT_LIMIT`, as per his suggestion, `DEBUG_PRINT_LIMIT` wont work in no env mode. I was thinking it for a moment, if we do like `defaultOptions`, then we need to export it. So was thinking, what your thoughts on that? Yes to me, thats a good way to support across all the env.

File: src/queries.js (hunk 1 above)
Comment (lgandecki) [reply src/queries.js]: I'm not against using pretty-format, I'm just trying to understand if there is a reason to not put it in package.json as a direct dependency. Then there is no need to fight around eslint complaining (rightly). And if jest-matcher-utils switch to something else internally and stop requiring pretty-format, our package will still work.

Comment (kentcdodds) [reply src/queries.js]: Agreed

Comment (antoaravinth) [reply src/queries.js]: @lgandecki Yes, I agree. Will make the changes. Thanks your comments!

**Comment by antoaravinth** [comment]
Ping, any thoughts on my last comment? Guess people missed it ;)

**Comment by antoaravinth** [comment]
Ping, any thoughts on my last comment? Guess people missed it ;)  cc @kentcdodds

**Comment by kentcdodds** [comment]
I have thoughts. Patience. Not all of us have time to address this within hours or days. We'll get to it when we can 😉

One thing you could do is rebase your branch and fix merge conflicts 😀

**Comment by antoaravinth** [comment]
@kentcdodds Oops! Sorry thought people had missed it. will do the rebase and update the branch.

File: src/__tests__/element-queries.js [hunk 5]
```diff
@@ -193,4 +193,30 @@ test('using jest helpers to check element class names', () => {
...
+    </div>`
+  const {getByTestId} = render(Hello)
+  process.env.DEBUG_PRINT_LIMIT = 10 // user should see `. . .`
+  expect(() => expect(getByTestId('not present')).toBeInTheDOM()).toThrowError()
+
+  const {getByLabelText} = render(Hello)
+  process.env.DEBUG_PRINT_LIMIT = 10000 // user shouldn't see `. . .`
+  expect(() =>
+    expect(getByLabelText('not present')).toBeInTheDOM(),
+  ).toThrowError()
```
Comment (kentcdodds) [inline src/__tests__/element-queries.js]: Make sure to reset `DEBUG_PRINT_LIMIT` to clean things up.

It should look something like this:

```javascript
test('test the debug helper prints the dom state here', () => {
  const originalDebugPrintLimit = process.env.DEBUG_PRINT_LIMIT

  // the test contents

  process.env.DEBUG_PRINT_LIMIT = originalDebugPrintLimit
})
```

That way it doesn't mess up other tests.

File: src/__tests__/element-queries.js [hunk 6]
```diff
@@ -193,4 +193,30 @@ test('using jest helpers to check element class names', () => {
...
+
+  const {getByText} = render(Large) // render large DOM which exceeds 7000 limit
+  expect(() => expect(getByText('not present')).toBeInTheDOM()).toThrowError()
+
+  const Hello = `<div data-testid="debugging" data-otherid="debugging">
+        Hello World!
+    </div>`
+  const {getByTestId} = render(Hello)
+  process.env.DEBUG_PRINT_LIMIT = 10 // user should see `. . .`
+  expect(() => expect(getByTestId('not present')).toBeInTheDOM()).toThrowError()
```
Comment (kentcdodds) [inline src/__tests__/element-queries.js]: This is a good assertion, but we should add another assertion to make sure the error includes the `...`

You can do this by adding a regex to the `toThrowError` call:

`.toThrowError(/\.\.\./)`

In the one below we should do the opposite


```javascript
.toThrowError(/^((?!\.\.\.).)*$/)
```

I'm _pretty_ sure that should verify that there is no `...` in the error message.

**Comment by sompylasar** [comment]
@kentcdodds Did we go with unchecked process.env for configuration?

**Comment by sompylasar** [comment]
> That said, anyone using this in the browser will be using a bundler so I'm pretty sure this'll be fine.

I thought about this, too. Although I'm not 100% sure.

@antoaravinth Could you please check if this holds true at least for Cypress? You can do so from `cypress-testing-library`: clone it, update `dom-testing-library` dependency, `npm run -s test:cypress` or `test:cypress:open`. Thanks!

**Comment by antoaravinth** [comment]
@sompylasar Yes it will. I have checked anyways:

```
Opening Cypress...
  (Tests Starting)


  dom-testing-library commands
    ✓ getByPlaceholderText (537ms)
    ✓ getByText (129ms)
    ✓ getByLabelText (495ms)
    ✓ getByAltText (134ms)
    ✓ getByTestId (134ms)


  5 passing (2s)


  (Tests Finished)

  - Tests:           5
  - Passes:          5
  - Failures:        0
  - Pending:         0
  - Duration:        2 seconds
  - Screenshots:     0
  - Video Recorded:  false
  - Cypress Version: 2.1.0


  (All Done)
```

**Comment by sompylasar** [comment]
@antoaravinth Awesome! :+1:

**Pull request merged**